internal/pkg/agent/application/actions/handlers/handler_helpers.go

```diff
@@ -0,0 +1,104 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package handlers
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/pkg/component"
+)
+
+type actionCoordinator interface {
+	State() coordinator.State
+	PerformAction(ctx context.Context, comp component.Component, unit component.Unit, name string, params map[string]interface{}) (map[string]interface{}, error)
+}
+
+type upgradeCoordinator interface {
+	actionCoordinator
+	Upgrade(ctx context.Context, version string, sourceURI string, action *fleetapi.ActionUpgrade, skipVerifyOverride bool, pgpBytes ...string) error
+}
+
+type performActionFunc func(context.Context, component.Component, component.Unit, string, map[string]interface{}) (map[string]interface{}, error)
+
+type dispatchableAction interface {
+	MarshalMap() (map[string]interface{}, error)
+	Type() string
+}
+
+func dispatchActionInParallel(ctx context.Context, log *logp.Logger, action dispatchableAction, comps []component.Component, units []component.Unit, performAction performActionFunc) error {
+	if action == nil {
+		return nil
+	}
+
+	// Deserialize the action into map[string]interface{} for dispatching over to the apps
+	params, err := action.MarshalMap()
+	if err != nil {
+		return err
+	}
+
+	actionType := action.Type()
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Iterate through found components and forward the UNENROLL action
```

Comment:
Misleading comment? This code doesn't look specific to UNENROLL

Reply:
Yes, was a result of refactoring the reusable function. Fixed.